kernel/arch/dreamcast/hardware/g2dma.c

```diff
@@ -0,0 +1,234 @@
+static ptr_t dma_cbdata[4];
+
+static int dma_init;
+
+/* G2 Bus DMA registers */
+static volatile g2_dma_reg_t * const extdma = (g2_dma_reg_t *)0xa05f7800;
+
+/* 
+    DMA Initiation Triggers:
+        - CPU_TRIGGER: Software-driven. (Setting enable and start to 1)
+        - HARDWARE_TRIGGER: Via AICA (DMA0) or expansion device.
+        - INTERRUPT_TRIGGER: Based on interrupt signals. 
+*/
+#define CPU_TRIGGER       0
+#define HARDWARE_TRIGGER  1
+#define INTERRUPT_TRIGGER 2
+
+/*  
+    For addresses, ensure bits 31-29 & 4-0 are '0' to avoid illegal interrupts. 
+    Only bits 28-5 are used for valid addresses. 
+*/
+#define MASK_ADDRESS 0x1fffffe0
+
+/*
+    Controls DMA initiation behavior after a DMA transfer completes:
+        0x00000000: Preserve the current DMA enable setting.
+        0x80000000: Reset the DMA enable setting to "0" after transfer.
+
+    OR '|' this value with length to set the size of DMA request.
+*/
+#define PRESERVE_ENABLED  0x00000000
+#define RESET_ENABLED     0x80000000
+
+/* 
+    Specifies system memory address range for G2-DMA across channels 0-3.
+    Overruns outside this range trigger G2-DMA overrun errors. Previous 
+    setting was 0x4659404f representing range 0x0C400000 - 0x0C4F0000. 
+*/
+#define SYS_MEM_SECURITY_CODE 0x4659
+#define DISABLE_SYS_MEM_PROTECTION (SYS_MEM_SECURITY_CODE << 16 | 0x007F)
+#define ENABLE_SYS_MEM_PROTECTION  (SYS_MEM_SECURITY_CODE << 16 | 0x7F00)
+
+/*  
+    Set the DS# (Data Strobe) timeout to 27 clock cycles for the external DMA. 
+    If data isn't ready for latching by this time, an interrupt will be 
+    triggered. Not sure why its 27 but can be changed later. Default value 
+    is 1023 cycles.
+*/
+#define DS_CYCLE_OVERRIDE  27
+
+/* Disable the DMA */
+static void dma_disable(uint32_t chn) {
+    extdma->dma[chn].enable = 0;
+    extdma->dma[chn].start = 0;
+}
+
+static void g2_dma_irq_hnd(uint32_t code) {
+    int chn = code - ASIC_EVT_G2_DMA0;
+
+    if(chn < 0 || chn > 3) {
+        dbglog(DBG_INFO, "g2_dma: Wrong channel received in g2_dma_irq_hnd");
+        chn = 1;
+    }
+
+    /* VP : changed the order of things so that we can chain dma calls */
+
+    /* Signal the calling thread to continue, if any. */
+    if(dma_blocking[chn]) {
+        sem_signal(&dma_done[chn]);
+        thd_schedule(1, 0);
+        dma_blocking[chn] = 0;
+    }
+
+    /* Call the callback, if any. */
+    if(dma_callback[chn]) {
+        dma_callback[chn](dma_cbdata[chn]);
+    }
+}
+
+int g2_dma_transfer(void * sh4, void * g2bus, uint32_t length, uint32_t block,
+                    g2_dma_callback_t callback, ptr_t cbdata,
+                    uint32_t dir, uint32_t mode, uint32_t g2chn, uint32_t sh4chn) {
+    /* No longer used but we keep then around for compatibility */
+    (void)mode;
+    (void)sh4chn;
+
+    if(g2chn > 3) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    /* Make sure we're not already DMA'ing */
+    if(extdma->dma[g2chn].start != 0) {
```

Comment:
Perhaps this should be the last of the error checks to give it the most time to get reset and to group 'input checks' together then the 'system check'.

Reply:
Another good point.  I just copied the check structure approach used in pvrdma.c.  We should probably update it there too.  Another PR...